ribs/archives/_unstructured_archive.py

```diff
@@ -0,0 +1,524 @@
+"""Contains the UnstructuredArchive class."""
+from functools import cached_property
+from typing import Any, Dict, List, Union
+
+import numpy as np
+
+from ribs._utils import (check_batch_shape, check_finite, check_shape,
+                         np_scalar, validate_batch, validate_single)
+from ribs.archives._archive_base import ArchiveBase
+from ribs.archives._transforms import (batch_entries_with_threshold,
+                                       compute_best_index,
+                                       compute_objective_sum,
+                                       single_entry_with_threshold)
+
+
+class UnstructuredArchive(ArchiveBase):
+    """An archive that adds new solutions based on their novelty.
+
+    This archive is described in `Lehman 2011
+    <https://www.cs.swarthmore.edu/~meeden/DevelopmentalRobotics/lehman_gecco11.pdf>`_.  # pylint: disable=C0301
+    If a solution is in a sparse area of metric space it is added
+    unconditionally. When a solution is in an overdense region it is added to
+    the archive only if its objective improves upon the nearest existing
```

Comment:
Kind of - the paper by Lehman I referenced includes an objective but it is actually not what I described here so I will change the code to more closely match the original work. I think this means a new transform or two since rather than just comparing objective values, they use a method more akin to non-dominating sort of three objectives:

1. novelty - sparsity in measure space
2. genotypic diversity - uniqueness of network topology
3. fitness - the number of nearby solutions with objective scores less than the solution

A solution is added iff it is better than solutions in its neighborhood on any objective and simultaneously no worse than its neighborhood on all other objectives. So a solution which improves novelty, but "dominates" fewer neighbors in fitness than each of those neighbors would not be added.

At this point, since all networks have essentially the same genotype we are only comparing objectives (1) and (3)

Reply:
Perhaps the best solution instead of doing the kNN search in `index_of` as I initially did, we *only* do it in a transform. The `index_of` would then just return the nearest neighbor already in the archive. It would still be a little expensive as we would be calculating distances twice (once in `index_of` and again in the transform) but cheaper than doing kNN twice.

Reply:
I see the confusion. Joel and Ken released two papers that year. The first was Novelty Search. The second was Novelty Search with Local Competition, which you linked above. Novelty Search is the one we're looking for: http://eplex.cs.ucf.edu/papers/lehman_ecj11.pdf

Reply:
Regarding kNN, I think what you have currently is almost the correct logic, and you should not need to modify transforms. Basically, if the solution is novel enough, then just assign it a new index in the archive, and the current add method should work just fine because it will never have to deal with solutions that are in the same index.

If you don't mind, I'd like to take a pass on this over the next few days; I think I have a good idea of how to do it.

Reply:
Novelty Search alone is certainly much simpler to implement than what I was stepping into. However, I think it would be nice to include both behaviors. The authors seemed to indicate local competition exceeded both novelty alone, and global competition.

My only concern with slotting a novel solution into a new index is how quickly the memory footprint might grow. Maybe you have an idea of how to do this efficiently.

I am curious to see what you come up with.

I am basically done implementing the NS with local competition. The Stanley and Lehman treat the problem as a MOEA where the objectives are fitness and genotypic variation. They take the NSGA-II approach but replace the "crowding distance" with "novelty" as the authors suggest. To that end I made the archive class able to specify additional objectives to include as part of the MOEA problem with the default behavior being single objective: maximizing fitness.

Also, to that end, I implemented some non-dominating sort algorithms... however I think covering NDS algorithms would expand the scope of the PyRibs project considerably. If including NS + local competition is desired, then it might be best to add a dependency on another project where NDS algorithms are in-scope. These are some projects I know of which cover NDS algorithms, PyMOO is the best that I'm aware of:

- [PyGAD](https://pygad.readthedocs.io/en/latest/) | [code](https://github.com/ahmedfgad/GeneticAlgorithmPython/blob/master/pygad/utils/nsga2.py)
  - this one only implements NDS in python, and only has a single implementation of NDS
- [PyMOO](https://pymoo.org/) | [code](https://github.com/anyoptimization/pymoo/tree/main/pymoo/util/nds)
  - this library has the best support for NDS since it has quite a few algorithms implemented, as well as Cython variants which speed up the algorithm considerably over pure python
- [Sferes2](https://sferes2.github.io/sferes2/) | [code](https://github.com/sferes2/sferes2/blob/master/sferes/ea/dom_sort.hpp)
  - this is a C++ library so including it would be a major hurdle, it only implements a single NDS algorithm, and hasn't been updated in nearly a decade... I included it simply because quite a few research papers and other projects use it or reference it.
- [Pygmo](https://esa.github.io/pygmo2/index.html) | [code](https://github.com/esa/pygmo2/blob/master/pygmo/core.cpp#L405)
  - this is a mixed C++/Python library which is a Python port of [Pagmo2](https://esa.github.io/pagmo2/). Its major drawback is that not only does it require the Pagmo2 C++ library be installed, but just a single NDS algorithm is implemented

Reply:
(Resolving; responded in general thread)